flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableMetaStoreFactory.java

```diff
@@ -140,12 +203,194 @@ private void alterPartition(
+                }
+                HiveStatsUtil.updateStats(accumulate(statsList), statistic);
+                return Optional.of(statistic);
+            } else {
+                // if the input format is neither OrcInputFormat nor MapredParquetInputFormat,
+                // we can't gather full statistic in current implementation.
+                // so return empty.
+                return Optional.empty();
+            }
+        }
+
+        /** List data files recursively. */
+        private List<FileStatus> listDataFileRecursively(FileSystem fileSystem, Path f)
+                throws IOException {
+            List<FileStatus> fileStatusList = new ArrayList<>();
+            for (FileStatus fileStatus : fileSystem.listStatus(f)) {
+                if (fileStatus.isDir() && !isStagingDir(fileStatus.getPath())) {
+                    fileStatusList.addAll(
+                            listDataFileRecursively(fileSystem, fileStatus.getPath()));
+                } else {
+                    if (isDataFile(fileStatus)) {
+                        fileStatusList.add(fileStatus);
+                    }
+                }
+            }
+            return fileStatusList;
+        }
+
+        private boolean isStagingDir(Path path) {
+            // in batch mode, the name for staging dir starts with '.staging_', see
+            // HiveTableSink#toStagingDir
+            // in stream mode, the stage dir is the partition/table location, but the staging files
+            // start with '.'
+            return path.getPath().startsWith(".");
+        }
+
+        private boolean isDataFile(FileStatus fileStatus) {
+            String fileName = fileStatus.getPath().getName();
+            return !fileName.startsWith(".")
+                    && !fileName.startsWith("_")
+                    && !fileName.equals(successFileName);
+        }
+
+        private Class<? extends InputFormat<?, ?>> getInputFormatClz(String inputFormatClz) {
+            try {
+                return (Class<? extends InputFormat<?, ?>>)
+                        Class.forName(
+                                inputFormatClz,
+                                true,
+                                Thread.currentThread().getContextClassLoader());
+            } catch (ClassNotFoundException e) {
+                throw new FlinkHiveException(
+                        String.format(
+                                "Unable to load the class of the input format %s.", inputFormatClz),
+                        e);
+            }
+        }
+
+        private Future<CatalogTableStatistics> submitStatsGatherTask(
+                Callable<CatalogTableStatistics> statsGatherTask) {
+            if (executorService == null) {
+                executorService =
+                        gatherStatsThreadNum == 1
+                                ? newDirectExecutorService()
+                                : Executors.newFixedThreadPool(3);
```

Comment:
I think the size of the fixed thread pool should not be a constant, it should use the user-configured size?

Reply:
Good catch. A mistake in here.